docs/src/references/mqtt-topic-guidelines.md

```diff
@@ -0,0 +1,640 @@
+
+Where `main` is used as an alias for the tedge device id.
+
+**Why have the `/device/` subtopic level?**
+
+To make a clear distinction between `device` data and `service` data.
+This distinction makes it easier to subscribe to all device only data excluding the service data
+with simple subscription filters like `tedge/main/device/#`.
+Without that device suffix, something like `tedge/main/#` would include the services' data as well.
+
+
+**Why have the `/service/` subtopic level and not have the <service-id> directly?**
+
+Primarily for future-proofing as `service` is a new kind of child abstraction that we've added now.
+If another abstraction is introduced in the future, say `plugin`, they can be namespaced under a `plugin/` subtopic.
+It also makes the distinction from `device` data clearer.
+
+### Telemetry
+
+For telemetry data, the topics would be grouped under a `telemetry/` sub-topic as follows:
+
+|Type|Example|
+|------|-------|
+|Measurements|`<target_prefix>/telemetry/measurements/[/<measurement-type>]`|
+|Events|`<target_prefix>/telemetry/events/<event-type>`|
+|Alarms|`<target_prefix>/telemetry/alarms/<alarm-type>`|
+
+The `measurement-type` is optional for `measurements`.
+The alarm `severity` which was earlier part the topic have been moved to the payload.
+
+**Examples**
+
+The following shows an example of the measurement telemetry topics for each of the target types:
+
+|Target|Topic structure|
+|------|-------|
+|tedge device|`tedge/main/device/telemetry/measurements`|
+|service on tedge device|`tedge/main/service/<service-id>/telemetry/measurements`|
+|child device|`tedge/<child-id>/device/telemetry/measurements`|
+|service on child device|`tedge/<child-id>/service/<service-id>/telemetry/measurements`|
+
+**Why have the redundant `/telemetry` subtopic level?**
+
+To have a clear separation from commands or other kinds of data that might get added in future.
+It simplifies subscriptions for "all telemetry data from a device" to just `tedge/<device-id>/+/telemetry/#`.
+Without that `telemetry` grouping level, doing such a subscription would be difficult as you'll have to subscribe to each of the following topics separately:
+
+* `tedge/<device-id>/+/measurements/#`
+* `tedge/<device-id>/+/events/#`
+* `tedge/<device-id>/+/alarms/#`
+
+In additional, this topic structure allow to combine the subscription of telemetry data as well as command data via the following MQTT wildcard topic:
+
+```
+tedge/<device-id>/+/#
+```
+
+Creating ACL rules is also simplified by allowing telemetry data to be controlled independently to the command topics.
+
+### Commands
+
+Similarly, all commands would be grouped under a `commands/` sub-topic as follows:
+
+|Target|Topic structure|
+|------|-------|
+|For requests|`tedge/<device-id>/<target-type>/commands/req/<operation-type>/<operation-specific-keys>...`|
+|For responses|`tedge/<device-id>/<target-type>/commands/res/<operation-type>/<operation-specific-keys>...`|
```

Comment:
Something unclear to me is how to distinguish the MQTT topics for **mapper<-> plugin** and **plugin<->child**.

Imagine the use case.
1. A tedge device receives cloud-specific MQTT firmware operation for child device.
2. Cloud mapper converts the message and publishes to the topic `tedge/child1/device/commands/req/device/firmware_update` ?
3. Firmware plugin receives the message, and publish the new message to `tedge/child1/device/commands/req/device/firmware_update`??? The same topic?

Reply:
>  how to distinguish the MQTT topics for mapper<-> plugin and plugin<->child.

Indeed an issue. Having dedicated topics for requests and responses implies that these topics have to be assigned each to a single entity (either a plugin/service or a child device) responsible of all these requests or responses. 

This is where I think we need to consider the benefits of using [MQTT-based state machine](https://github.com/didier-wenzek/tedge-mqtt-state-machine).

In this proposal, there is no separate topics for requests and responses but a topic dedicated to an operation instance. All the actors of operation (the mapper, the plugin and the child device) share this topic to listen to the operation progress and to participate when this is their turn (e.g. the mapper creates an init state, the plugin downloads the firmware, the child installs the firmware once locally available, ...).

Reply:
> Firmware plugin receives the message, and publish the new message tedge/child1/device/commands/req/device/firmware_update??? The same topic?

That's not what I had in mind. The mapper will covert a c8y operation to tedge command as you mentioned in step.2. Currently the `c8y-firmware-plugin` is doing this mapping as well and doing other things like maintaining the firmware cache etc as well. But, when we convert the `c8y-firmware-plugin` into a `tedge-firmware-plugin`, all these cloud specific responsibilities would be moved to the c8y-mapper instead. And when that happens, the mapped tedge commands can be directly consumed by the child devices. In fact, there won't be a `tedge-firmware-plugin` until we start supporting firmware management for the main device.

It's much more clearer in the case of the upcoming `tedge-config-plugin` and `tedge-log-plugin`. When the mapper sends the config requests for the main device, these plugins will consume those and read/update the config filesor log files on the main device. For the child devices, they consume the requests sent by the mapper directly, as the `tedge-xyz-plugin` need not intercept that.

>  All the actors of operation (the mapper, the plugin and the child device) share this topic to listen to the operation progress and to participate when this is their turn (e.g. the mapper creates an init state, the plugin downloads the firmware, the child installs the firmware once locally available, ...).

Yeah, that's another way to model it as well. In my head, even downloading the artefacts is a responsibility of the mapper, as these artefacts are mostly downloaded from the cloud itself. So, if the mapper manages downloads and uploads to the cloud, there's no need for the another plugin on the device to intercept those requests for child devices and reforward that.

But, we should probably discuss this shift, in the light of the other work converting all the c8y plugins to tedge plugins as well.

Reply:
> It's much more clearer in the case of the upcoming tedge-config-plugin and tedge-log-plugin. When the mapper sends the config requests for the main device, these plugins will consume those and read/update the config filesor log files on the main device. For the child devices, they consume the requests sent by the mapper directly, as the tedge-xyz-plugin need not intercept that.

I think your design is well valid only for upload. For any operations requiring download, the timing of publishing a MQTT message ("commands/req") is **after** downloading is done.
You mean mapper should decide to download when it receives a cloud specific operation? It is more than message conversion.

But I partly agree that mapper should be in charge of upload/download as authentication is cloud-specific. My proposal is to use MQTT request/response from plugin for upload/download.

We can discuss this in coming weeks. For operations, MQTT topics and mapper-plugin-agent model are connected discussion point.

Reply:
This conversation is about outdated content.